scripts/helper_functions.sh

```diff
@@ -93,47 +96,205 @@ export YellowBoldText='\033[1;33m'      # Yellow
+
+    temp_file=$(mktemp)
+    http_code=$(curl https://api.steamcmd.net/v1/info/2394010 --output "$temp_file" --silent --location --write-out "%{http_code}")
+    TARGET_MANIFEST=$(grep -Po '"2394012".*"gid": "\d+"' <"$temp_file" | sed -r 's/.*("[0-9]+")$/\1/')
+
+    CURRENT_MANIFEST=$(awk '/manifest/{count++} count==2 {print $2; exit}' /palworld/steamapps/appmanifest_2394010.acf)
+    rm "$temp_file"
+
+    if [ "$http_code" -ne 200 ]; then
+        LogError "There was a problem reaching the Steam api. Unable to check for updates!"
+        DiscordMessage "There was a problem reaching the Steam api. Unable to check for updates!" "failure"
+        return 2
+    fi
+
+    if [ -z "$TARGET_MANIFEST" ]; then
+        LogError "The server response does not contain the expected BuildID. Unable to check for updates!"
+        DiscordMessage "Steam servers response does not contain the expected BuildID. Unable to check for updates!" "failure"
+        return 2
+    fi
+
+    if [ "$CURRENT_MANIFEST" != "$TARGET_MANIFEST" ]; then
+        LogInfo "An Update Is Available."
+        LogInfo "Current Version: $CURRENT_MANIFEST"
+        LogInfo "Latest Version: $TARGET_MANIFEST."
+    return 0
+    fi
+
+    LogSuccess "The Server is up to date!"
+    return 1
+
+}
+
+InstallServer() {
+    DiscordMessage "${DISCORD_PRE_UPDATE_BOOT_MESSAGE}" "in-progress"
+    /home/steam/steamcmd/steamcmd.sh +@sSteamCmdForcePlatformType linux +@sSteamCmdForcePlatformBitness 64 +force_install_dir "/palworld" +login anonymous +app_update 2394010 validate  +quit
+    DiscordMessage "${DISCORD_POST_UPDATE_BOOT_MESSAGE}" "success"
+}
+
+# Returns 0 if game is installed
+# Returns 1 if game is not installed
+IsInstalled() {
+    if  [ -e /palworld/PalServer.sh ] && [ -e /palworld/steamapps/appmanifest_2394010.acf ]; then
+        return 0
+    fi
+    return 1
+}
+
+# Saves the server
+# Returns 0 if it saves
+# Returns 1 if it is not able to save
+save_server() {
+    local return_val=0
+    if ! RCON save; then
+        return_val=1
+    fi
+    return "$return_val"
+}
+
+# Saves then shutdowns the server
+# Returns 0 if it is shutdown
+# Returns 1 if it is not able to be shutdown
+shutdown_server() {
+    local return_val=0
+    # Do not shutdown if not able to save
+    if save_server; then
+        if ! RCON "DoExit"; then
+            return_val=1
+        fi
+    else
+        return_val=1
+    fi
+    return "$return_val"
+}
+
+# Given a message this will broadcast in game
+# Since RCON does not support spaces this will replace all spaces with underscores
+# Returns 0 on success
+# Returns 1 if not able to broadcast
+broadcast_command() {
+    local return_val=0
+    # Replaces spaces with underscore
+    local message="${1// /_}"
+    if [[ $TEXT = *[![:ascii:]]* ]]; then
+        LogWarn "Broadcast message contains non-ascii characters: \"${message}\""
+    fi
+    if ! RCON "broadcast ${message}"; then
+        return_val=1
+    fi
```

Comment:
@Luatan do you think we should always check if the message being broadcast contains non-ascii characters or only when the message is broadcast?
```suggestion
    if RCON "broadcast ${message}"; then
        # If message is broadcast then log if it contains non-ascii characters
        if [[ $TEXT = *[![:ascii:]]* ]]; then
            LogWarn "Broadcast message contains non-ascii characters: \"${message}\""
        fi
    else
        return_val=1
    fi
```

Reply:
I think we should check before we actually send it. I just tested it a bit. It seems like it does broadcast, but it also gives me an error:
```
docker exec -it palworld-server rcon-cli "broadcast 🙌"
Broadcasted: ð�
cli: execute: rcon: read tcp 127.0.0.1:45320->127.0.0.1:25575: i/o timeout
```
Also there won't be anything "useful" broadcasted anyway.